src/api/blobs/reader.rs

```diff
@@ -0,0 +1,327 @@
+            let guard = &mut this.state;
+            match std::mem::take(guard) {
+                ReaderState::Idle { position } => {
+                    // todo: read until next page boundary instead of fixed size
+                    let len = buf.remaining() as u64;
+                    let end = position.checked_add(len).ok_or_else(|| {
+                        io::Error::new(ErrorKind::InvalidInput, "Position overflow when reading")
+                    })?;
+                    // start the export op for the entire size of the buffer, and convert to a stream
+                    let stream = this
+                        .blobs
+                        .export_ranges(this.options.hash, position..end)
+                        .stream();
+                    position1 = Some(position);
+                    *guard = ReaderState::Reading {
+                        position,
+                        op: Box::pin(stream),
+                    };
+                }
+                ReaderState::Reading { position, mut op } => {
+                    let position1 = position1.get_or_insert(position);
+                    match op.poll_next(cx) {
+                        Poll::Ready(Some(ExportRangesItem::Size(_))) => {
+                            *guard = ReaderState::Reading { position, op };
+                        }
+                        Poll::Ready(Some(ExportRangesItem::Data(data))) => {
+                            if data.offset != *position1 {
+                                break Poll::Ready(Err(io::Error::other(
+                                    "Data offset does not match expected position",
+                                )));
+                            }
+                            buf.put_slice(&data.data);
+                            // update just local position1, not the position in the state.
+                            *position1 =
+                                position1
+                                    .checked_add(data.data.len() as u64)
+                                    .ok_or_else(|| {
+                                        io::Error::new(ErrorKind::InvalidInput, "Position overflow")
+                                    })?;
+                            *guard = ReaderState::Reading { position, op };
+                        }
+                        Poll::Ready(Some(ExportRangesItem::Error(err))) => {
+                            *guard = ReaderState::Idle { position };
+                            break Poll::Ready(Err(io::Error::other(format!(
+                                "Error reading data: {err}"
+                            ))));
+                        }
+                        Poll::Ready(None) => {
+                            // done with the stream, go back in idle.
+                            *guard = ReaderState::Idle {
+                                position: *position1,
+                            };
+                            break Poll::Ready(Ok(()));
```

Comment:
might be easier to read if you use `return `  for these? maybe that's just me

Reply:
I had break so I could log the value before returning during debugging.